In [1]:
import requests
import os
from dotenv import load_dotenv
import collections


In [2]:
API_KEY = ''

In [3]:
params = dict(key=API_KEY, date='2022-04-06')
res = requests.get('https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X',params)

print(res)

<Response [200]>


In [4]:
json = res.json()
json.keys()

dict_keys(['dataset'])

In [5]:
params = dict(start_date='2017-01-01', end_date='2017-12-31', api_key=API_KEY)
data_2017 = requests.get('https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X',params)

In [6]:
data_json = data_2017.json()
data_json = dict(data_json['dataset'])

These are your tasks for this mini project:

Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
Convert the returned JSON object into a Python dictionary.
Calculate what the highest and lowest opening prices were for the stock in this period.
What was the largest change in any one day (based on High and Low price)?
What was the largest change between any two days (based on Closing Price)?
What was the average daily trading volume during this year?
(Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [7]:
type(data_json)

dict

In [8]:
data_json.keys()

dict_keys(['id', 'dataset_code', 'database_code', 'name', 'description', 'refreshed_at', 'newest_available_date', 'oldest_available_date', 'column_names', 'frequency', 'type', 'premium', 'limit', 'transform', 'column_index', 'start_date', 'end_date', 'data', 'collapse', 'order', 'database_id'])

In [9]:
data_json['column_names']

['Date',
 'Open',
 'High',
 'Low',
 'Close',
 'Change',
 'Traded Volume',
 'Turnover',
 'Last Price of the Day',
 'Daily Traded Units',
 'Daily Turnover']

In [10]:
#Calculate what the highest and lowest opening prices were for the stock in this period.
open_values = {}
for values in data_json['data']:
    if values[0] != None and values[1] != None:
        date = values[0]
        open_price = values[1]
        open_values[date] = open_price
maximum = max(open_values, key=open_values.get)
print(maximum, open_values[maximum])
minimum = min(open_values, key=open_values.get)
print(minimum, open_values[minimum])  

2017-12-14 53.11
2017-01-24 34.0


### The highest opening price was 53.11 on 2017-12-14
### The lowest opening price was 34.0 on 2017-01-24

In [11]:
#What was the largest change in any one day (based on High and Low price)?
change_values = {}
for values in data_json['data']:
    #if values[0] != None and values[1] != None:
    date = values[0]
    change_price = values[2] - values[3]
    change_price = float("{:.2f}".format(change_price))
    change_values[date] = change_price
maximum_change = max(change_values, key=change_values.get)  # Just use 'min' instead of 'max' for minimum.
print(maximum_change, change_values[maximum_change])

2017-05-11 2.81


### The largest change in any one day was 2.81$ on 2017-05-11

In [12]:
#What was the largest change between any two days (based on Closing Price)?
closing_values = {}
for values in data_json['data']:
    #if values[0] != None and values[1] != None:
    date = values[0]
    closing_price = values[4]
    #change_price = float("{:.2f}".format(change_price))
    closing_values[date] = closing_price

In [13]:
values = list(closing_values.values())
xdiff = [values[n]-values[n-1] for n in range(1,len(values))]
max_diff = float("{:.2f}".format(max(xdiff)))
max_diff

2.56

### The largest change between any two days based on Closing price: 2.56$

In [14]:
#What was the average daily trading volume during this year?
trading_volume = {}
for values in data_json['data']:
    date = values[0]
    trading_vol = values[6]
    trading_volume[date] = trading_vol

In [15]:
sum_val = sum(trading_volume.values())
len_dict = len(trading_volume)
average_tr = sum_val/len_dict
average_tr = float("{:.2f}".format(average_tr))
average_tr

89124.34

### The average daily trading volume during the year was 89124.34$

In [16]:
#(Optional) What was the median trading volume during this year. 
#(Note: you may need to implement your own function for calculating the median.)
def median_number(dictionary):
    sorted_val = sorted(dictionary.values())
    length = len(sorted_val)
    if length%2 != 0:
        median_index = int((length+1)/2)
        print(sorted_val[median_index])
    else:
        first_n = length/2
        second_n = (length/2)+1
        median_index = int((first_n+second_n)/2)
        print(sorted_val[median_index])
median_number(trading_volume)

76600.0


### The median trading volume: 76600.0$